### PR TITLE
Convert setup.cfg to pyproject.toml

### LINKED ISSUE #1276: Consolidate requirements in a single place
Right now, when we change requirements, we need to change them in three places:

 - `setup.cfg`
 - `requirements/*.txt`
 - `requirements/environment.yml`

As I have empirically demonstrated 😬, there's a good chance that we'll end up forgetting to update them in all three places and end up with inconsistencies, even with the comments that state the other places where things need to be changed.  

We should consolidate requirements in a single place, if there is a reasonable way to go about it.  The problem is that I don't know if there currently is a reasonable way to do this.  For example, there is information in `environment.yml` that is not contained in the other two places (i.e. which channels to use and whether packages get installed via those channels or via `pip`).  

The previous major changes to the requirements in #789 were a huge and much-appreciated improvement over what was there before, and I really like the way that the requirements are split up in each of the `setup.cfg` and the `requirements/*.txt` files, so it'd be really helpful to keep a structure like if there is a way to consolidate them.

### LINKED ISSUE #1680: Move from `setup.py` and `setup.cfg` to `pyproject.toml`
The Python community has been converging on using [`pyproject.toml`](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/). The old way of doing it with `setup.py` and `setup.cfg` is [now considered legacy](https://pip.pypa.io/en/stable/reference/build-system/setup-py/).  The Python Packaging User Guide's tutorial on [packaging Python projects](https://packaging.python.org/en/latest/tutorials/packaging-projects/) now includes `pyproject.toml` and doesn't include `setup.cfg` or `setup.py`.

We should move away from using our [`setup.cfg`](https://github.com/PlasmaPy/PlasmaPy/blob/main/setup.cfg) for configuration and packaging stuff when we have the option to, and instead use our [`pyproject.toml`](https://github.com/PlasmaPy/PlasmaPy/blob/main/pyproject.toml) file.  We should also drop our [`setup.py`](https://github.com/PlasmaPy/PlasmaPy/blob/main/setup.py`) and have `pyproject.toml` handle those responsibilities.

We are also currently using `setup.cfg` for configuring `flake8`, `pytest`, `coverage`, and `codespell`.  We should move the configurations out of `setup.cfg` and into `pyproject.toml` or elsewhere, probably splitting this up into multiple smaller pull requests to make each individual chunk more manageable.

This isn't high priority, but it would be good to do this in the next year or two.

### PR DESCRIPTION
**Pull request by StanczakDominik**
## Description

I converted setup.cfg to pyproject.toml using https://ini2toml.readthedocs.io/, and validated it using https://pypi.org/project/validate-pyproject/.

I also had to bump pytest to 6.0 so it supports pyproject.toml.

I also managed to rewrite readthedocs' config so that it sources the dependencies from `pyproject.toml`. This means we no longer need to store the separate files of `requirements.txt` type.

I sidestepped flake8 not supporting pyproject.toml via moving its settings to... `tox.ini`, which it, for some reason, accepts. Since that's a file we're keeping anyway, it looked like the cleanest approach.

Also, I made two more noteable changes:
- I removed the "`extras`" and "`all`" "extra"  pip dependencies. `extras` was literally a subset of developer tools, so there was no need to keep it. `all`  depended on a quirk of setup.cfg and pyproject.toml doesn't reimplement that fortunate glitch. We could tinker with `cog` or we could require a recent `setuptools/pip` as per https://hynek.me/articles/python-recursive-optional-dependencies/, but I thought it's not worth it, so I just limited the extras to `tests`, `docs` and `developer`, `developer` manually including the former two (and being shortened to `dev`). I guess for development we could ask people to update setuptools + pip, and then we could make the upper two subsets. Does that make sense?
- I removed `environment.yml`. In recent years it's been reduced to setting the python version and naming the environment and did fairly little.

# Remaining todo

- [x] ensure proper version generation
- [x] adjust documentation
- [ ] does this work with binder?
- [x] investigate Erik issues with setuptools-scm and editable installs
- [x] investigate whether one can do pip install . for just deps

## Motivation and context

Mostly #1680; but I've been the victim of `setup.py` based packages at work due to how `poetry` interacts with them` (tldr: arbitrary python code execution just to solve dependencies is terrible) and I wanted us to be part of the change I want to see in the world.

## Related issues

Closes #1680. Closes #1276. Initially meant to get tested by #1742, but that one turns out to be blocked.

### CODE REVIEW

**Comment by StanczakDominik** [comment]
I have a commit in place to deal with the coverage issue, but don't want to stop the doc build which finally finished after several trial-and-error attempts. The trick there was updating pip and setuptools at the exact right moment before the package gets built - which is basically the fix from https://github.com/lief-project/LIEF/issues/108#issuecomment-337246904.

**Comment by StanczakDominik** [comment]
- [ ] https://github.com/astropy/astropy/pull/10774 may be a solution to the versioning issue

**Comment by StanczakDominik** [comment]
I think this is going to do it. cc @rocco8773, please test out whether you can reproduce your setuptools-scm problem with an `--editable` install.

**Comment by StanczakDominik** [comment]
It should work now but it looks like you're at a conference or smth

**Comment by StanczakDominik** [comment]
So the main difference now is that we're checking the SCM version CONDITIONALLY on being in an editable install, because of the manifest.in file

**Comment by StanczakDominik** [comment]
I'm 90% sure this is going to fix @rocco8773's docs issue, so I'd be in favor of reviewing and merging this soon-ish.

File: tox.ini
Comment (namurphy) [inline tox.ini:202]:
```suggestion
    RST306
    # TC003 to allow long messages to be specified outside the exception class
    TC003
```
Adding in a `tryceratops` error code to ignore (from #1782).

**Review by namurphy — COMMENTED** [review]
It's looking good!  I'm going to merge a suggestion to fix the linter issues.

File: changelog/1758.trivial.rst
Comment (namurphy) [inline changelog/1758.trivial.rst:1]:
```suggestion
Replaced :file:`setup.py` with :file:`pyproject.toml` (see :pep:`518` and :pep:`621`).
```
Sphinx!

**Comment by StanczakDominik** [comment]
- [ ] revert back to two-line version for convenience

**Comment by StanczakDominik** [comment]
Seems to be done! @namurphy can I get a reapproval on this? I'm trying to be better about virtualenvs on my fresh Arch install and trying to use [Hatch](https://hatch.pypa.io/latest/environment/#dependencies, but it's complaining about our `pyproject.toml` on main being ancient 😁 

I briefly checked binder and it seems like that **doesn't** work... but it doesn't work on main atm, either! So I'll fix that in a separate PR.

**Comment by StanczakDominik** [comment]
Oh crud, setup.cfg went back during the merge. Hold off please. Fixing.

**Review by namurphy — COMMENTED** [review]
Thank you again for working on this! 

I can't remember...did we decide anything about the requirements files? Were we planning on keeping them for the time being?  (Admittedly I kind of want to delete them, see if their absence breaks anything without good workarounds, and then put them back only if needed.  That might be a bit risky though...)

Following up on one of our earlier conversations about this, it turns out that `pip` has a [`--target`](https://pip.pypa.io/en/latest/cli/pip_install/#cmdoption-t) flag that can be used to specify the directory that the package should be installed in.  I *think* that means that doing a `pip install -e .[all] --target=/some/other/directory` would create an editable installation without polluting the clone of the repo with a bunch of installation files.  I don't know if it would handle the dev version number correctly.

Apart from the question about the `requirements` files, this looks to be in good shape.  Since there are a few PRs by new contributors open right now, it may be better to wait until they're merged before merging this (just in case of any unforeseen complications).

Thank you again!

File: .readthedocs.yml
Comment (StanczakDominik) [reply .readthedocs.yml:14]: Ooooh, I think forgetting about them for this PR was what what reminded me of this bug! :D

File: docs/contributing/doc_guide.rst
Comment (StanczakDominik) [reply docs/contributing/doc_guide.rst:398]: `.flake8rc`, I suppose!

File: docs/install.rst
Comment (StanczakDominik) [reply docs/install.rst:197]: We absolutely would, I just forgot them here!

I'm of two minds about including them - on one hand, we don't lose much by mentioning this (two characters...), but on the other hand, it's sort of a special case and if you're alternative enough to use zshell I'm pretty sure you're expected to know your shell's special cases!

Thanks for point it out, I'll add them here.

File: plasmapy/_dev/scm_version.py
Comment (StanczakDominik) [reply plasmapy/_dev/scm_version.py:10]: I would absolutely change it right now, but it's early morning, I'm not yet hydrated nor caffeinated, I dare not at the moment. Let's just get it in.

File: docs/contributing/install_dev.rst
Comment (StanczakDominik) [reply docs/contributing/install_dev.rst:106]: Yeah, we absolutely should! Thanks for catching this.

Comment (StanczakDominik) [reply docs/contributing/install_dev.rst:106]: Or maybe not, grep for "Next, setup the development version of PlasmaPy which you just cloned". It's there, maybe not with the best possible flow, but it's there, and I don't want to break the structure we have now.

**Comment by StanczakDominik** [comment]
> I can't remember...did we decide anything about the requirements files? Were we planning on keeping them for the time being? (Admittedly I kind of want to delete them, see if their absence breaks anything major, and then put them back only if needed. That might be a tad risky though...)

Literally second-to-last telecon (since I missed the last one), I had this exchange with Erik:

> E: did you remember to bring back requirements files?
> D: no

Which is why if you look [just above](https://github.com/PlasmaPy/PlasmaPy/pull/1758#issuecomment-1331165647), you'll find:

> * [x]  revert back to two-line version for convenience
> * [x]  **bring back requirements.txt file**

So yeah. I completely agree about wanting to remove them, though. Separate PR, I guess.

**Comment by StanczakDominik** [comment]
> Following up on one of our earlier conversations about this, it turns out that `pip` has a [`--target`](https://pip.pypa.io/en/latest/cli/pip_install/#cmdoption-t) flag that can be used to specify the directory that the package should be installed in. I _think_ that means that doing a `pip install -e .[all] --target=/some/other/directory` would create an editable installation without polluting the clone of the repo with a bunch of installation files. I don't know if it would handle the dev version number correctly.

I think you *might* be right, but 50-50 on whether it'd actually work without breaking... I'd have to study the interaction between the two flags in detail.

**Comment by StanczakDominik** [comment]
pre-commit.ci autofix

**Comment by StanczakDominik** [comment]
But... They are in there... Aren't they?

**Comment by namurphy** [comment]
It looks like `requirements.txt` and `requirements/extras.txt` are still being deleted, though it looks like what's in `extras.txt` was moved elsewhere.

If we keep the `requirements` files then we will probably want to regenerate (in the style of Doctor Who) `test_requirements.py`, but I'm okay with deleting it here for now.  I'm not sure what we'll want to do with `environment.yml`, but I'm okay with deleting that for now too.

**Comment by StanczakDominik** [comment]
Eh.

You know, I still just don't see a convincing use case for them. I think I'll just delete them completely (in the morning) and I'll happily revert that removal once a need for them is precisely defined, but... I don't know, I just don't see the need for now.

**Comment by namurphy** [comment]
> Eh.
> 
> You know, I still just don't see a convincing use case for them. I think I'll just delete them completely (in the morning) and I'll happily revert that removal once a need for them is precisely defined, but... I don't know, I just don't see the need for now.

Yeah...my vote would be to delete them, try out the changes for a while, and then revert them if we discover that there is a compelling use case for them.

**Comment by StanczakDominik** [comment]
@namurphy I've removed them, would quite appreciate a re-review!

I'll just add a coverage-ignore comment for the "plasmapy is not installed section".

File: plasmapy/__init__.py
Comment (namurphy) [inline plasmapy/__init__.py:52]:
```suggestion
    __version__ = "0.0.0"
```
Should this be `__version__` here?

**Review by namurphy — APPROVED** [review]
LGTM!  I have one minor suggestion for a possible typo.  Thank you again so much for doing this!

File: plasmapy/__init__.py
Comment (StanczakDominik) [reply plasmapy/__init__.py:52]: It certainly should! Good catch, thanks!

**Comment by StanczakDominik** [comment]
Thanks for the review, extremely happy to finally get this in and unblock the other PRs!

**Pull request merged**
